runtime/src/main/java/org/corfudb/util/serializer/KeyDynamicProtobufSerializer.java

```diff
@@ -0,0 +1,135 @@
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.DynamicMessage;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.ByteBufOutputStream;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.CorfuStoreMetadata;
+import org.corfudb.runtime.collections.CorfuDynamicKey;
+import org.corfudb.runtime.collections.OpaqueCorfuDynamicRecord;
+import org.corfudb.runtime.exceptions.SerializerException;
+
+import java.io.IOException;
+
+@Slf4j
+public class KeyDynamicProtobufSerializer extends DynamicProtobufSerializer {
+
+    public KeyDynamicProtobufSerializer(CorfuRuntime corfuRuntime) {
+        super(corfuRuntime);
+    }
+
+    /**
+     * Deserialize an object from a given byte buffer.
+     *
+     * @param b The bytebuf to deserialize.
+     * @return The deserialized object.
+     */
+    @Override
+    public Object deserialize(ByteBuf b, CorfuRuntime rt) {
+
+        try (ByteBufInputStream bbis = new ByteBufInputStream(b)) {
+            ProtobufSerializer.MessageType type = ProtobufSerializer.MessageType.valueOf(bbis.readInt());
+            int size = bbis.readInt();
+            byte[] data = new byte[size];
+            bbis.readFully(data);
+            CorfuStoreMetadata.Record corfuRecord = CorfuStoreMetadata.Record.parseFrom(data);
+            Any payload = corfuRecord.getPayload();
+
+            String fullMessageName = getFullMessageName(payload);
+            if (!messagesFdProtoNameMap.containsKey(fullMessageName)) {
+                log.error("messagesFdProtoNameMap doesn't contain the message type {} of payload {}." +
+                                "Please check if the related table is properly opened with correct schema.",
+                        fullMessageName, payload);
+                log.error("messagesFdProtoNameMap keySet is {}", messagesFdProtoNameMap.keySet());
+            }
+
+            if (type.equals(ProtobufSerializer.MessageType.KEY)) {
+                Descriptors.FileDescriptor valueFileDescriptor
+                        = getDescriptor(messagesFdProtoNameMap.get(fullMessageName));
+                Descriptors.Descriptor valueDescriptor = valueFileDescriptor.findMessageTypeByName(getMessageName(payload));
+                DynamicMessage value = DynamicMessage.parseFrom(valueDescriptor, payload.getValue());
+                return new CorfuDynamicKey(payload.getTypeUrl(), value);
+            }
+
+            String metadataTypeUrl = null;
+            ByteString metaDataByteString = null;
+            if (corfuRecord.hasMetadata()) {
+                Any anyMetadata = corfuRecord.getMetadata();
+                metadataTypeUrl = anyMetadata.getTypeUrl();
+                metaDataByteString = anyMetadata.getValue();
+
+                String fullMetadataMessageName = getFullMessageName(anyMetadata);
+                if (!messagesFdProtoNameMap.containsKey(fullMetadataMessageName)) {
```

Comment:
This would still hit a Deserialization exception if there is a problem with the type of the record right?
Instead can we just directly return the byte[] as the type itself symmetrically across serialize() and deserialize() so that if something goes wrong, the DyanmicProtobufSerializer does not come in the middle?

Reply:
This only logs the error but now fail the compactor. Logging such error can assist the debugging to find out when such issue happened for the first time.